Introduction to Examples + Compositor Explanation (#1812)
opendp/opendp: docs/source/getting-started/examples_intro.rst

> Getting Started With Examples
> ==========

Reviewer: Make title reflect the directory name? and add `:orphan:`
```suggestion
:orphan:
:nosearch:

Working with Tabular Data
=========================
```

2. **Sample Utility:** The public micro-data is a sample of the more thorough research microdata available through scientific use files. Methods developed with the public microdata are designed to also work on the scientific files. Researchers can request access to the full dataset through Eurostat. 

Dataset Description 
-------------------

The data is organized by year and quarter for each nation in the European Union. For this tutorial, we sampled 200,000 individuals from the public microdata of France across all study years. 

The data is protected using traditional statistical disclosure control methods such as global recoding, local suppression, and addition of noise. 

Learn more about the dataset `here <https://ec.europa.eu/eurostat/web/microdata/public-microdata/labour-force-survey>`_! The `user guide <https://ec.europa.eu/eurostat/documents/1978984/6037342/EULFS-Database-UserGuide.pdf>`_ may also be helpful to learn more about the different variables.

Core Variables 
--------------
There are many variables available. The examples will focus on the following key variables: 

.. list-table:: 
   :header-rows: 1

   * - Variable
     - Definition
     - Coding

>    * - `SEX`
>      - Sex
>      - 1: Male \n 2: Female
>    * - `AGE`
>      - Age of the Individual During the Reference Week
>      - Single Years
>    * - `ILOSTAT`
>      - Labour Status During the Reference Week
>      - 1: Did any work for pay or profit during the reference week - one hour or more (including family workers but excluding conscripts on compulsory military or community service) \n 2: Was not working but had a job or business from which he/she was absent during the reference week (including family workers but excluding conscripts on compulsory military or community service) \n 3: Was not working because of lay-off \n 4: Was a conscript on compulsory military or community service \n 5: Other (15 years or more) who neither worked nor had a job or business during the reference week \n 9: Not applicable (child less than 15 years old)
>    * - `HWUSUAL`
>      - Number of Hours Per Week Usually Worked
>      - 00: Usual hours cannot be given because hours worked vary considerably from week to week or from month to month \n 01-98: Number of hours usually worked in the main job \n 99: Not applicable (WSTATOR=3-5,9) \n blank: No answer
>    * - `QUARTER`
>      - Fixed Reference Quarter
>      - Single Quarter
>    * - `YEAR`
>      - Fixed Reference Year

Reviewer: - `\n` isn't special for sphinx. Replacing with semicolons.
- double backticks for code formatting here.
- Leave off `WSTATOR=3-5,9` if we don't define it.

```suggestion
   * - ``SEX``
     - Sex
     - 1: Male; 2: Female
   * - ``AGE``
     - Age of the Individual During the Reference Week
     - Single Years
   * - ``ILOSTAT``
     - Labour Status During the Reference Week
     - 1: Did any work for pay or profit during the reference week - one hour or more (including family workers but excluding conscripts on compulsory military or community service); 2: Was not working but had a job or business from which he/she was absent during the reference week (including family workers but excluding conscripts on compulsory military or community service); 3: Was not working because of lay-off; 4: Was a conscript on compulsory military or community service; 5: Other (15 years or more) who neither worked nor had a job or business during the reference week; 9: Not applicable (child less than 15 years old)
   * - ``HWUSUAL``
     - Number of Hours Per Week Usually Worked
     - 00: Usual hours cannot be given because hours worked vary considerably from week to week or from month to month; 01-98: Number of hours usually worked in the main job; 99: Not applicable; blank: No answer
   * - ``QUARTER``
     - Fixed Reference Quarter
     - Single Quarter
   * - ``YEAR``
     - Fixed Reference Year
```

-------------------
The compositor is the foundation of our differentially private queries. It essentially takes in our data and our specifications for the queries that we would like to run. At this point, we won't be directly referencing our data again and we could theoretically delete it! 

.. code-block:: python

   context = dp.Context.compositor(
       data=df,
       privacy_unit=dp.unit_of(contributions=36),
       privacy_loss=dp.loss_of(epsilon=1.0),
       split_evenly_over=10,
       margins={
           ("SEX", ): dp.Margin(max_partition_length=60_000_000),
           ("AGE", ): dp.Margin(max_partition_length=60_000_000),
           ("ILOSTAT", ): dp.Margin(max_partition_length=60_000_000),
           ("HWUSUAL", ): dp.Margin(max_partition_length=60_000_000),
           ("YEAR", ): dp.Margin(max_partition_length=60_000_000, max_partition_contributions=4),
           ("QUARTER", ): dp.Margin(max_partition_length=60_000_000, max_partition_contributions=13),
           ("YEAR", "QUARTER",): dp.Margin(max_partition_length=60_000_000, max_partition_contributions=1),
           (): dp.Margin(max_partition_length=60_000_000),
       },
   )

**Parameters**:

> * *privacy_unit:* How many rows each individual or entity of interest contributes to our data frame. In this case, we are analyzing the data from the first quarter of 2008 and each individual is represented once. Hence, the privacy unit is 1. If we were to analyze the dataset across quarters and years, each individual would contribute (number of quarters * number of years) times.

Reviewer: > the privacy unit is 1

conflicts with

> `dp.unit_of(contributions=36)`

Author: fixed

> * *privacy_loss:* This parameter determines how much privacy we want to preserve. If ε is small, we will have more privacy but worse data accuracy. ε can range from 0 to infinity, but 1 is usually a standard. For now, we won't specify an ε.

Reviewer: > For now, we won't specify an ε.

conflicts with

> `dp.loss_of(epsilon=1.0)`

Author: fixed - we will specify

> * *split_evenly_over:* This is the number of queries you want to distribute your privacy loss over.

Reviewer: Maybe say that we're using 10 just to explore the API, but in the final version we'd pick it more carefully?

* *margins:* Margins capture the variables of interest in your analysis. This can include variables that you may want to group by or apply differential privacy techniques to. 

>     * *max_partition_length:* The upper bound on how many records (individuals in this case) can be in one partition. This is generally an upper bound on the population represented in your dataset if you do not know the size of your dataset. In this case, the population of France was about 60 million in 2004 so that's our maximum partition length. Source: `World Bank <https://datatopics.worldbank.org/world-development-indicators/>`. 

Reviewer: ```suggestion
    * *max_partition_length:* The upper bound on how many records (individuals in this case) can be in one partition. If you do not know the size of your dataset, this can be an upper bound on the population represented in your dataset. The population of France was about 60 million in 2004 so that's our maximum partition length. Source: `World Bank <https://datatopics.worldbank.org/world-development-indicators/>`_. 
```

>     * *max_partition_contributions:* The number of contributions each individual can have per grouping. Since each individual is represented once for a particular quarter and year, they are represented 13 times for each quarter since there are 13 in the dataset and 4 times each year since there are 4 quarters within a year. 

Reviewer: ```suggestion
    * *max_partition_contributions:* The number of contributions each individual can have per grouping. Since each individual is represented once for a particular quarter and year, they are represented 13 times for each quarter since there are 13 years in the dataset and 4 times each year since there are 4 quarters within a year. 
```

Reviewer: Add spaces between these lines for RST formatting, and double ticks around the names.

> Particular examples will require additional parameters, and the compositor will change slightly. You can read more about the parameters `here <https://docs.opendp.org/en/nightly/api/python/opendp.context.html#opendp.context.Context.compositor>`. 

Reviewer: ```suggestion
Particular examples will require additional parameters, and the compositor will change slightly.
See :py:func:`opendp.context.Context.compositor` for more information.
```

This way, if we changed the name of that function, the docs build would immediately catch the bad link, rather than only catching it after nightly was updated.

The `toctree` can be added at the bottom.
